### PR TITLE
Implement calicoctl patch

### PR DESCRIPTION
**Pull request by syscll**
## Description

This is a base implementation of a new `calicoctl patch` command as requested in #2047 

The end goal is to best match the implementation of `kubectl patch`. However, to start, the functionality of this PR will be limited to JSON data only and will perform a strategic merge patch as described [here](https://kubernetes.io/docs/tasks/run-application/update-api-object-kubectl-patch/#use-a-strategic-merge-patch-to-update-a-deployment).

## Todos
- [x] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
calicoctl now supports patching resources using `calicoctl patch`
```

### CODE REVIEW

File: calicoctl/commands/patch.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+  # Partially update a node using a json merge patch.
+  calicoctl patch node node-0 --patch '{"spec":{"unschedulable":true}}' --type json
+
+Options:
+  -h --help                  Show this screen.
+  -p --patch=<PATCH>         Spec to use to patch the resource.  If set
+                             to "-" loads from stdin.
+  -t --type=<TYPE>           Format of patch type:
+                                json        JSON Patch, RFC 6902 (not yet implemented)
+								merge       JSON Merge Patch, RFC 7386 (not yet implemented)
```
Comment (caseydavenport) [inline calicoctl/commands/patch.go]: Indentation here maybe looks a bit funny

Comment (syscll) [reply calicoctl/commands/patch.go]: Fixed

**Comment by syscll** [comment]
Will add some unit and integration tests, as well as perform some manual testing.

File: calicoctl/commands/patch.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+	"fmt"
+	"strings"
+
+	"github.com/docopt/docopt-go"
+	"github.com/projectcalico/calicoctl/calicoctl/commands/constants"
+	log "github.com/sirupsen/logrus"
+)
+
+func Patch(args []string) error {
+	// TODO: write coherent description.
```
Comment (caseydavenport) [inline calicoctl/commands/patch.go]: Is this TODO still valid?

Comment (syscll) [reply calicoctl/commands/patch.go]: Ah, yes, I suppose it was more of a reminder to get it proofread by someone else. So if you're happy, I'll remove the TODO

Comment (caseydavenport) [reply calicoctl/commands/patch.go]: Yeah, this looks fine to me!

**Comment by syscll** [comment]
What's the procedure around release notes and docs? Do I just need to add a `patch` section to the `calicoctl` reference for v3.10?

**Comment by caseydavenport** [comment]
@syscll yep - I've added a release note to the description which will get picked up during the next release.

We should add new documentation to the `master/` directory of the Calico repo (since that will become v3.10). We'll definitely want the reference changes, we might also want to update some of our other docs which have users do a get->modify->update to just use patch instead, since it's way simpler.

**Comment by caseydavenport** [comment]
Let's make sure to squash into one commit before merging.

**Pull request merged**